Fix #4751: Update work manager to 2.7.0 and enable it on S+ devices

theMr17 opened the pull request:
<!-- READ ME FIRST: Please fill in the explanation section below and check off every point from the Essential Checklist! -->
## Explanation
<!--
  - Explain what your PR does. If this PR fixes an existing bug, please include
  - "Fixes #bugnum:" in the explanation so that GitHub can auto-close the issue
  - when this PR is merged.
  -->

Fixes #4751

This PR updates `WorkManager` from version 2.4.0 to 2.7.0.

The upgrade ensures compatibility with Android 12 (API level 31) and above, where background work scheduling requires newer `WorkManager` APIs. Additionally, since the project already uses Kotlin 1.6, the version bump aligns well with the current dependency ecosystem and does not introduce any major breaking changes.

The `androidx.work.impl.WorkManagerInitializer` entry was removed from the main manifest because `WorkManager` ≥ 2.6.0 no longer includes that class (initialization moved to AndroidX Startup). The `InitializationProvider` handling was added in their respective test manifests instead of the main manifest because test modules have different application IDs, dependency classpaths, and merge behavior--so each test target needs a manifest overlay that matches its classpath and avoids manifest merge conflicts.

|Device|Logs|
|--|--|
|Older|<img width="1822" height="354" alt="image" src="https://github.com/user-attachments/assets/024274c1-49d6-464f-895a-f66fc2522306" />|
|S+|<img width="1821" height="403" alt="image" src="https://github.com/user-attachments/assets/0d02fb9b-75ce-41fa-998c-f9cb5efeb9ce" />|

## Essential Checklist
<!-- Please tick the relevant boxes by putting an "x" in them. -->
- [x] The PR title and explanation each start with "Fix #bugnum: " (If this PR fixes part of an issue, prefix the title with "Fix part of #bugnum: ...".)
- [x] Any changes to [scripts/assets](https://github.com/oppia/oppia-android/tree/develop/scripts/assets) files have their rationale included in the PR explanation.
- [x] The PR follows the [style guide](https://github.com/oppia/oppia-android/wiki/Coding-style-guide).
- [x] The PR does not contain any unnecessary code changes from Android Studio ([reference](https://github.com/oppia/oppia-android/wiki/Guidance-on-submitting-a-PR#undo-unnecessary-changes)).
- [x] The PR is made from a branch that's **not** called "develop" and is up-to-date with "develop".
- [x] The PR is **assigned** to the appropriate reviewers ([reference](https://github.com/oppia/oppia-android/wiki/Guidance-on-submitting-a-PR#clarification-regarding-assignees-and-reviewers-section)).

theMr17 committed:
Update work manager version to 2.7.0 and enable it on S+ devices

theMr17 committed:
Remove unused import

theMr17 committed:
Update maven_dependencies.textproto

theMr17 committed:
Replace the conflicting android:authorities

theMr17 committed:
Replace the conflicting android:authorities in generic modules

theMr17 committed:
Remove WorkManagerInitializer from app/AndroidManifest

theMr17 commented:
@adhiamboperes PTAL.

theMr17 committed:
Remove TODO

adhiamboperes commented on app/src/main/AndroidManifest.xml:
```diff
@@ -364,11 +364,6 @@
     <activity
       android:name=".app.testing.TextInputLayoutBindingAdaptersTestActivity"
       android:theme="@style/OppiaThemeWithoutActionBar" />
-    <provider
-      android:name="androidx.work.impl.WorkManagerInitializer"
-      android:authorities="${applicationId}.workmanager-init"
-      android:exported="false"
-      tools:node="remove" />
```
Why was this removed from the main manifest file, but added to all other files? -- This sounds important to document in the PR description.

theMr17 replied:
I have updated the PR description with this explanation.

adhiamboperes replied:
Sg, thanks.

adhiamboperes reviewed (commented):
Thanks for the meticulous work, @theMr17! I have one question inline.

Additionally, could you please do manual tests on an S+ device and an older device to verify that logs are stil being uploaded correctly? You can do this by:
1. Enabling analyticscollection via the `AndroidManifest`
2. Checking log cat logs to see whether Firebase upload is actually succeeding.

theMr17 committed:
Move android:authorities override to main Manifest file

theMr17 committed:
Revert "Move android:authorities override to main Manifest file"

This reverts commit e941e43340a9b8a220f1d01b3e625d83c24fa531.

theMr17 commented:
Hey @adhiamboperes, I've addressed your comment and added screenshots confirming that logs are still being uploaded correctly.

I do have one more question, though. Since `WorkManager` no longer requires `WorkManagerInitializer` and instead relies on the AndroidX Startup library, the following manifest entry should technically be removable:

https://github.com/oppia/oppia-android/blob/c47a03918b65505926f04146d858a0b256336654/instrumentation/src/java/AndroidManifest.xml#L218-L222

I'm unsure how to properly verify that the instrumentation module continues to work as expected after removing it. Do you have any suggestions on how to validate this safely?

adhiamboperes commented:
> oppia-android/instrumentation/src/java/AndroidManifest.xml

@theMr17, I believe that you can attempt to run `instrumentation/src/javatests/org/oppia/android/instrumentation/player/ExplorationPlayerTest.kt` and see if it will build with either configuration. This is a test-only module.

adhiamboperes reviewed (commented):
Other than the open question, this PR LGTM.

BenHenning reviewed (approved):
Thanks @theMr17! This is fantastic, and looks like a great and clearly explained fix. Really happy to see this get addressed.

Do we know if this fixes #5442?

theMr17 committed:
Remove provider from instrumentation Manifest

theMr17 commented:
Thanks for the review @BenHenning!

Re: Its difficult to verify if #5442 (or even #5452) got fixed. As both are production issues and have unclear repro steps. We might need to rely on future Play Console logs to confirm if this PR fixes those as well (which ideally should be the case).

theMr17 commented:
> I do have one more question, though. Since `WorkManager` no longer requires `WorkManagerInitializer` and instead relies on the AndroidX Startup library, the following manifest entry should technically be removable:

@adhiamboperes this is now done, PTAL.
